R/kmeans-clustering-method.R

```diff
@@ -0,0 +1,528 @@
+      # Find the elbow using the method described in
+      # stackoverflow.com/a/2022348/633251
+      # but translated to R (see above).
+      
+      # Add an index to argument values for easy plotting
+      DF <- data.frame(x = 1:length(wss), y = wss)
+      fit <- lm(y ~ x, DF[c(1,nrow(DF)),]) # 2 point 'fit'
+      m <- coef(fit)[2]
+      b <- coef(fit)[1]
+      
+      # Check to make sure the data is concave as described
+      # in the documentation, as arbitrary trends could give
+      # misleading answers.  The following approach simply
+      # checks to make sure all values are either above or
+      # below the reference line.  This allows the values
+      # to vary quite a bit and still return an answer.
+      
+      concave <- FALSE
+      use <- 2:(nrow(DF) - 1)
+      refpts <- m*DF$x[use] + b
+      if (all(refpts > DF$y[use]) | all(refpts < DF$y[use])) concave <- TRUE
+      if (!concave) stop("Your curve doesn't appear to be concave")
+      
+      # Calculate the orthogonal distances
+      use <- 2:(nrow(DF) - 1)
+      elbowd <- private$distancePointLine(DF$x[use], DF$y[use], coef(fit)[2], coef(fit)[1])
+      DF$dist <- c(NA, elbowd, NA) # first & last points don't have a distance
+      
+      private$vnum <- which.max(DF$dist)
+      
+      if (isTRUE(self$params$debug)) {
+        print('found the optimal number of clusters...')
+        print(private$vnum)
+      }
+  
+    }, # end of findElbow
+    
+    
+    # Override: run k-means algorithm
+    buildClusters = function() {
+      
+      ## Use principle components if usePrinComp is TRUE and the numOfPrinComp is given
+      if (self$params$usePrinComp == TRUE && is.null(self$params$numOfPrinComp)) {
+        stop("Principle components are not created because the numbder of principle components 
+             is not dfined.")
+      } else if (self$params$usePrinComp == TRUE && !is.null(self$params$numOfPrinComp)) {
+        df <- private$prin.comp
+      } else {
+        df <- self$params$df
+      }
+      
+      ## If the numOfClusters is not given, use the one generated by findElbow()
+      if (!is.null(self$params$numOfClusters)) {
+        numOfClusters <- self$params$numOfClusters 
+      } else {
+        numOfClusters <- private$vnum
+      }
+      
+      # Standarize the variables
+      df.stand <- private$scaledf
+      # K-Means
+      private$kmeans.fit <- kmeans(df.stand, numOfClusters)
+      # Save the centers 
+      private$centers <- private$kmeans.fit[["centers"]]
+      private$cluster <- private$kmeans.fit[["cluster"]]
+      #print(kmeans.fit)
+      #return(invisible(private$kmeans.fit))
+    },
+    
+    # Generate confusion matrix
+    calculateConfusion = function() {
+      # generate a confusion matrix of clusters and labels
+      k.means.fit <- self$getKmeansfit()
+      clusters <- k.means.fit$cluster
+      #labels <- super$getLabelColVal()
```

Comment:
Dead code? 

Also, do you mind 

1) adding a line break to make this section more readable and

2) Capitalizing all comments?